### ISSUE TITLE
swift: Use Content-Type from directory listing to avoid lookups on 0 byte files in listings

### ISSUE THREAD

**Issue by xgiovio** [ncw/rclone#208]
Hello, an important feature for rclone could be : missing only.
The directories listing is very fast so find the missing files on remote side should be simple.

Usage scenario : rclone copy "local/path" "remote:container/path" --missing-only

with that option all mtimes, size and md5 checks are skipped

What do you think?

**Comment by ncw** [comment]
All the remotes give size on directory listing anyway so speed-wise this would be the same as `--size-only`.

**Comment by xgiovio** [comment]
i don't know why you say that the --size-only has the same speed. It isn't true!
Using cloudberry trial or swift commadnline i can obtain the listing of all my 5 tb in 30 sec and there are 700k files. 
using rclone with --size-only it took 45 minutes last night only for 40k files.

You are wrong when you say that the speed is the same. Please check you code or do some tests!

**Comment by ncw** [comment]
If you would like to to reproduce your results then please post the command lines you used - thanks!

**Comment by ncw** [comment]
Here is an example showing that a `--size-only` sync is quick with a swift backend.

```
$ rclone --dry-run --size-only sync local swift:Images
2015/11/09 20:44:52 Swift container Images: Building file list
2015/11/09 20:45:05 Swift container Images: Waiting for checks to finish
2015/11/09 20:45:06 Swift container Images: Waiting for transfers to finish
2015/11/09 20:45:06 Waiting for deletions to finish

Transferred:            0 Bytes (   0.00 kByte/s)
Errors:                 0
Checks:             58159
Transferred:          404
Elapsed time:       14.2s
```

And without `--size-only`

```
$ rclone --dry-run sync /home/ncw.irishsea/Images swift:Images2015/11/09 20:47:19 Swift container Images: Building file list
[snip]
2015/11/09 20:53:59 Swift container Images: Waiting for transfers to finish
2015/11/09 20:53:59 Waiting for deletions to finish

Transferred:            0 Bytes (   0.00 kByte/s)
Errors:                 0
Checks:             58159
Transferred:          404
Elapsed time:     6m40.1s
```

**Comment by xgiovio** [comment]
I have the same commands but with copy and not sync. I will try with sync but i wouldn't delete any files

**Comment by ncw** [comment]
Can you post the actual commands with output  like I did above.  You can use `--dry-run` like I did so no transfers or deletions actually happen. Doing it with `copy` is fine.

Could you also post the result of running `rclone size` on the container too please (along with times).

**Comment by xgiovio** [comment]
same problem.
> rclone rclone --dry-run --size-only sync "I:/newworks" "hubic:default/newworks"
2015/11/09 23:59:39 Swift container default path newworks/: Building file list
2015/11/10 00:00:38 
Transferred:            0 Bytes (   0.00 kByte/s)
Errors:                 0
Checks:                 0
Transferred:            0
Elapsed time:      1m2.9s

2015/11/10 00:01:38 
Transferred:            0 Bytes (   0.00 kByte/s)
Errors:                 0
Checks:                 0
Transferred:            0
Elapsed time:      2m2.9s

2015/11/10 00:02:38 
Transferred:            0 Bytes (   0.00 kByte/s)
Errors:                 0
Checks:                 0
Transferred:            0
Elapsed time:      3m2.9s

2015/11/10 00:03:38 
Transferred:            0 Bytes (   0.00 kByte/s)
Errors:                 0
Checks:                 0
Transferred:            0
Elapsed time:      4m2.9s

2015/11/10 00:04:38 
Transferred:            0 Bytes (   0.00 kByte/s)
Errors:                 0
Checks:                 0
Transferred:            0
Elapsed time:      5m2.9s

2015/11/10 00:05:38 
Transferred:            0 Bytes (   0.00 kByte/s)
Errors:                 0
Checks:                 0
Transferred:            0
Elapsed time:      6m2.9s

2015/11/10 00:06:38 
Transferred:            0 Bytes (   0.00 kByte/s)
Errors:                 0
Checks:                 0
Transferred:            0
Elapsed time:      7m2.9s

2015/11/10 00:07:38 
Transferred:            0 Bytes (   0.00 kByte/s)
Errors:                 0
Checks:                 0
Transferred:            0
Elapsed time:      8m2.9s

2015/11/10 00:08:38 
Transferred:            0 Bytes (   0.00 kByte/s)
Errors:                 0
Checks:                 0
Transferred:            0
Elapsed time:      9m2.9s

2015/11/10 00:09:38 
Transferred:            0 Bytes (   0.00 kByte/s)
Errors:                 0
Checks:                 0
Transferred:            0
Elapsed time:     10m2.9s

2015/11/10 00:10:38 
Transferred:            0 Bytes (   0.00 kByte/s)
Errors:                 0
Checks:                 0
Transferred:            0
Elapsed time:     11m2.9s

2015/11/10 00:11:38 
Transferred:            0 Bytes (   0.00 kByte/s)
Errors:                 0
Checks:                 0
Transferred:            0
Elapsed time:     12m2.9s

2015/11/10 00:12:38 
Transferred:            0 Bytes (   0.00 kByte/s)
Errors:                 0
Checks:                 0
Transferred:            0
Elapsed time:     13m2.9s

2015/11/10 00:13:38 
Transferred:            0 Bytes (   0.00 kByte/s)
Errors:                 0
Checks:                 0
Transferred:            0
Elapsed time:     14m2.9s

**Comment by ncw** [comment]
Can you try that with `--dump-headers` and see if you can work out what is happening?

**Comment by ncw** [comment]
Can you also try `rclone size "hubic:default/newworks"` please? And post the results. Thanks!

**Comment by xgiovio** [comment]
i'll do the tests in the following minutes.
However, yesterday nught i set up a little python script. It does exactly what i wanted.
1) create a local file list with sizes excluding folders
2) create a remote local list with sizes excludin folders
3) checking and correcting size of 0byte files if they have the manifest
4) give results

I can check and compare 300k files in less than a minute. So it's definitely an rclone issue.

Here the code. Sorry, no comments. I wrote it fast
>https://gist.githubusercontent.com/xgiovio/9e3eed6032c91efa064e/raw/57e33908391a8fcc7a4a3ceeba82d45bbf11c764/test.py

**Comment by xgiovio** [comment]
>Can you try that with --dump-headers and see if you can work out what is happening?

done, it send http request on each folder. I think they are 0 files folder. it should skip these checking the content type application/folder returned by directory listing

>Can you also try rclone size "hubic:default/newworks" please? And post the results. Thanks!

Done. Exactly the same situazion. It's slow because using --dump-headers i can see http requests on each folder.

I think you should only filters these 0bytes folder like the bug on past days on folders created by hubic

**Comment by ncw** [comment]
OK I think we are getting there!  What you are saying is that 0 byte files that hubic uses to mark directores are causing an HTTP request.

I can see why that is, and yes, I'm ignoring the content type in the listing.

I should be able to fix that relatively easily.

**Comment by xgiovio** [comment]
yes perfect. it can be a good bug fix for 1.25 milestone. (;

**Comment by xgiovio** [comment]
any news on the release? i would like to try the fix on the nas. could you send me a 1.25 compiled for arm? thank you

**Comment by ncw** [comment]
This fix has gone live in v1.25

Please test it and if you find any problems then open a new issue.

Thanks

Nick
